### ISSUE TITLE
FR - Implement Harvester using wix heat.

### ISSUE THREAD

**Issue by GilesBathgate** [oleg-shilo/wixsharp#695]
I've implemented a project harvester that uses wix heat as a backend and was wondering if this is something you would like to include in WixSharp. The implementation and example usage are in the 3 files of this Gist:

https://gist.github.com/GilesBathgate/02ec989cec836c097f35ea3517b5c31c

**Comment by oleg-shilo** [comment]
Thank you Gilles, will definitely have a look.

**Comment by GilesBathgate** [comment]
Ideally, the syntax would be different from the example in the Gist. I was thinking something like this would fit in better with the existing framework: 

```c#
var project = new ManagedProject(
                            new Dir(@"%programfiles%"/My Company/My Application",
                                    new ProjectReference(@"path/to/project.csproj")));
```

**Comment by oleg-shilo** [comment]
Sorry for the delay. Only now I fond the time to process your proposal.

This is how I understood it:
- _Heat_ is a WiX tool that crawls through the directory structure and generates WXS required for further building the MSI file.
- Your _Harvester_ is a solution for injecting the _Heat_ generated WXS in the WixSharp project structure so the MSI file can be built with WixSharp project.
 
If this is accurate then I do question the whole concept. Of course correct me if I misunderstood your solution.

WixSharp already has native support for aggregating the artifacts from the arbitrary folder ("ReleaseFolder" sample):

```C#
var project =
    new Project("MyProduct",
        new Dir(@"%ProgramFiles%\My Company\My Product",
            new Files(@"..\Release Folder\Release\*.*", f => !f.EndsWith(".obj") && !f.EndsWith(".pdb"))
. . .
```

It even more accurate version that only processes the top level files of the dir ("WildCard files" sample): 
```C#
var project =
    new Project("MyProduct",
        new Dir(@"%ProgramFiles%\MyCompany\MyProduct",
            new DirFiles(@"Release\Bin\*.*"),
            new Dir("GlobalResources",
                new DirFiles(@"Release\Bin\GlobalResources\*.*")),
            new Dir("Images",
                new DirFiles(@"Release\Bin\Images\*.*")),
            new ExeFileShortcut("Uninstall MyProduct", "[System64Folder]msiexec.exe", "/x [ProductCode]")));

project.UI = WUI.WixUI_FeatureTree;
project.GUID = new Guid("{AC19C6E4-9724-4e90-8AC7-6E69B4AB7562}");
```

Thus I am not sure what problem will be solved by bringing "Harvester" in the picture.

**Comment by GilesBathgate** [comment]
It's just intended as a neater and cleaner way of picking up only the outputs of the projects (*.vbproj, *.csproj), in particular, it creates all the subdirectories for localization satellite assemblies, and includes files marked as "content" of those projects.

The idea would be that the final implementation you would be able to do:

```c#
var project =
    new Project("MyProduct",
        new Dir(@"%ProgramFiles%\My Company\My Product",
            new ProjectReference(@"Path\To\Project.csproj")
```

You are right that you can do all of this manually, but that could be error-prone, and dosen't use the wix SDK in it's intended to be. I agree that the wixtoolset overcomplicated in many other places, but I thought the harvesting part worked quite well. It made migration of our wix based project to wix# much easier.

**Comment by GilesBathgate** [comment]
Here is an example of the (perhaps overly complex) content it generates:

```xml
<?xml version="1.0" encoding="utf-8"?>
<Wix xmlns="http://schemas.microsoft.com/wix/2006/wi">
    <Fragment>
        <DirectoryRef Id="INSTALLDIR">
            <Component Id="cmp977757D825BDFA8E13D00FE897C0D02C" Guid="*">
                <File Id="fil15AAB542F216052166BF4C549567EC20" Source="$(var.ProjectUI.TargetDir)\ProjectUI.dll" />
            </Component>
        </DirectoryRef>
    </Fragment>
    <Fragment>
        <ComponentGroup Id="ProjectUI.Binaries">
            <ComponentRef Id="cmp977757D825BDFA8E13D00FE897C0D02C" />
        </ComponentGroup>
    </Fragment>
    <Fragment>
        <DirectoryRef Id="INSTALLDIR">
            <Directory Id="dir4E151D1906C26C1D3F0BA781E2E5E52B" Name="Resources">
                <Component Id="cmpFB5B188AD5DDAEF7BBF0D105C5381BFB" Guid="*">
                    <File Id="fil9FD3B39049B43D9B59737B9F591193D1" Source="$(var.ProjectUI.ProjectDir)\Resources\Close_30px.png" />
                </Component>
                <Component Id="cmp538FBB267FF6AC56500A96680850B874" Guid="*">
                    <File Id="filEF073162CC00DBDB369AE41C0536227F" Source="$(var.ProjectUI.ProjectDir)\Resources\UnknownCategory.png" />
                </Component>
                <Component Id="cmp5AC8E0F446C37E62CA48C212C060518F" Guid="*">
                    <File Id="fil86639F4E92024D30210660074264EF30" Source="$(var.ProjectUI.ProjectDir)\Resources\Contract_100px.png" />
                </Component>
            </Directory>
        </DirectoryRef>
    </Fragment>
    <Fragment>
        <ComponentGroup Id="ProjectUI.Content">
            <ComponentRef Id="cmpFB5B188AD5DDAEF7BBF0D105C5381BFB" />
            <ComponentRef Id="cmp538FBB267FF6AC56500A96680850B874" />
            <ComponentRef Id="cmp5AC8E0F446C37E62CA48C212C060518F" />
        </ComponentGroup>
    </Fragment>
    <Fragment>
        <DirectoryRef Id="INSTALLDIR">
            <Directory Id="dirD96061F0F879E688F36ED851EDBBCB50" Name="ja-JP">
                <Component Id="cmp08A9139CC21509476F3A4CF44813071B" Guid="*">
                    <File Id="fil3726C07B02D0FC4CEEB952219151A0E4" Source="$(var.ProjectUI.TargetDir)\ja-JP\ProjectUI.resources.dll" />
                </Component>
            </Directory>
            <Directory Id="dirD6F68F3C436839749C82658ED3D9CD79" Name="gsw-LI">
                <Component Id="cmpDD73659FBE63F5AEA584BB7F937A33D9" Guid="*">
                    <File Id="fil2E4DCA2F4288B843DD6D2A5E65E7E43F" Source="$(var.ProjectUI.TargetDir)\gsw-LI\ProjectUI.resources.dll" />
                </Component>
            </Directory>
            <Directory Id="dir60673ABFAF1496D9DE4AB48F1E6008DE" Name="zh-Hans">
                <Component Id="cmp5F795DFD7EFDF691DC1B2FC82D58D858" Guid="*">
                    <File Id="filF9BC976CC1779E04A1FCCF9475EF6B06" Source="$(var.ProjectUI.TargetDir)\zh-Hans\ProjectUI.resources.dll" />
                </Component>
            </Directory>
        </DirectoryRef>
    </Fragment>
    <Fragment>
        <ComponentGroup Id="ProjectUI.Satellites">
            <ComponentRef Id="cmp08A9139CC21509476F3A4CF44813071B" />
            <ComponentRef Id="cmpDD73659FBE63F5AEA584BB7F937A33D9" />
            <ComponentRef Id="cmp5F795DFD7EFDF691DC1B2FC82D58D858" />
        </ComponentGroup>
    </Fragment>
</Wix>
```

**Comment by oleg-shilo** [comment]
I am not sure it is a cleaner way.
This is the proposed approach:

```C#
var project =
    new Project("MyProduct",
        new Dir(@"%ProgramFiles%\My Company\My Product",
            new ProjectReference(@"Path\To\Project.csproj")));
```

and this is what WixSharp already can do:
```C#
var project =
    new Project("MyProduct",
        new Dir(@"%ProgramFiles%\My Company\My Product",
            new Files(@"Path\To\Project\bin\Release\*.*")));
```
Seams like the second one is even cleaner.
If you are concern that it can pick wrong files then a simple class in the user or WixSharp codebase can do the trick:
```C#
class ProjectOutputFiles : Files
{
    public ProjectOutputFiles(string projectDir) 
        : base(projectDir.PathJoin("*.*"), 
               f => x.EndsWith(".exe") ||
                    x.EndsWith(".dll") ||
                    x.EndsWith(".config")) { }
} 
```
I have no problems with `ProjectOutputFiles` being a part of WixSharp.

> You are right that you can do all of this manually, but that could be error-prone...

Why do you say "manually"? `Files(...)` does it automatically. There is no any manual step involved. You specify what files you want to collect and they will be aggregated for you by the engine. And, opposite to _Heat_, _WixSharp_ does it in a FULLY controllable way. Thus it is actually no more but arguably less  "error-prone" than heat-based approach. 

> and dosen't use the wix SDK in it's intended to be.

Actually it's not the case. WixSharp had never had "the use of WiX SDK" as objective. I regret if WixSharp documentation or samples gave you that impression. The objective is to deliver to users features for building deployment solutions. And WiX is only a tool for achieving this purpose even if its name is reflected in the WixSharp name :)

And unfortunately WiX is not the easiest tool to deal with.  

As an example, consider a long standing [bug in Burn](https://github.com/wixtoolset/issues/issues/4921/), WiX own bootstrapping technology built on top of MSI. The bug prevents ANY msi authoring solution (not only WixSharp) from displaying _EmbeddedUI_ from the install session. The bug was reported (by me) about 5 years ago and it is still not fixed. Even though it affects the fundamentals of MSI. :( I even had difficult time to convince WiX team to accept the report.

Eventually WixSharp users were so fed up with this so one of them has contributed WixSharp extension for building NSIS bootstrappers instead. 

Thus if I have a choice to minimize dependency on any external tools I would definitely consider it as a preferred option.

**Comment by GilesBathgate** [comment]
>new Files(@"Path\To\Project\bin\Release\*.*")));

The output folder `bin\Release` is configurable. While I agree that it's probably sensible to use the visual studio defaults for our project we didn't. 

> a simple [ProjectOutputFiles] class in the user or WixSharp codebase can do the trick

Yeah that's good. But `.exe` `.config` and `.dll` are not the only outputs that would be harvested by `content`, `satellites`, and `binaries`

>WixSharp does it in a FULLY controllable way.

Well yeah, that's probably what I meant by manually. Certainly, that is a good thing if that is what you want, and it was by no means my intention to suggest a replacement for doing it based on files if that is what is wanted. The idea of this harvester is to allow users to do it based on projects, which users might be used to if coming from legacy wix.

>Thus if I have a choice to minimize dependency on any external tools.

That's cool, I didn't realise this was an objective.

**Comment by GilesBathgate** [comment]
Ultimately, I wrote this because it was useful for our usage of Wix#, and I wanted to contribute back to the open-source community. I am not that worried about it the extend of wanting a debate to justify it. It was more of a take it or leave it kind of thing which is why I submitted it as a Gist as opposed to a merge request.

**Comment by oleg-shilo** [comment]
Gilles, I only gave you such an extensive response because I truly appreciate your contributions and did not want to limit my answer to a simple response "it's inconsistent with Wix# development paradigm". Particularly after your recent involvement in developing Wix#! It would be just plain rude. :)

I still want to process your feedback and incorporate your proposal in one or another way. Even if I do not favor _heat_ approach. I see a few opportunities here. One is that I have shared and another one as a static method of `Files`, which parses the VS project for output dir and collects the assets:
```C#
new Dir(@"%ProgramFiles%\My Company\My Product",
    Files.FromProjectOutput(@"Path\To\Project.csproj")));
```
Though I would need to better understand the meaning of "content, satellites, and binaries". Can you please elaborate on what assets do qualify for being content, satellites, and binaries?

===================

And a few points about objectives :)

No, "minimize dependency" is not an objective neither. It is also an technical quality of the solution that I would break if I see enough benefits in doing so. 

When I created Wix# my objective was to free SW devs (as myself) from the pain of dealing directly from an extremely convoluted MSI authoring. I saw how WiX, after successfully solving tooling and source controlling problem of MSI, has felt short of addressing the flawed MSI authoring model/process. 

I can talk abut this a lot :) I am not sure you have time for that. But if you do, have a look at my interview that I gave to InfoQ ~5 years ago (https://www.infoq.com/articles/WixSharp/). 

Sorry for shamelessly referring to it but Jonathan (the interviewer) has asked very right questions and in the result that interview is the best highlight on "What is Wix# objective?".

**Comment by GilesBathgate** [comment]
Cool. I didn't intend to come across as flippant. On the point about your objectives, I agree. If only all software could be simple to use! For me, Wix# certainly meets that objective.

However, sometimes complexity is inherited from other frameworks and systems that underly them.

>which parses the VS project for output dir and collects the assets

I don't know whether the references to the files are  actually in the .vbproj files directly or created dynamically by some global msbuild `.targets` file.

Regarding that though I've just been discussing a use case with one of my colleagues and was pleased to find that our harvester works with his solution. Essentially we want to generate satellite assemblies on the fly and creating them as project outputs which heat picks up through msbuild "magic"

In summary (I think) the .csproj references a `Generator.targets` file which builds a dll dynamically, and then adds that dll to an MSBuild ProjectOutputGroupOutput <target> 
wixheat then picks up that dll and includes it in the wixsharp setup :confused: 

Now while this is a very contrived example, it just worked without any further changes to our migrated setup project.

**Comment by oleg-shilo** [comment]
I decided to give your Harvester a try. My thoughts where that parsing the .csproj may be OK but to find out the output produced by *.targets is close to impossible. It may be even calling a native dll at compile time. So I decided that I would keep your Harverster as is and also implement Harverster inspired `class ProjectOutputFiles`.

This is what I **was** thinking...

But after trying to invoke Harvester I realized that I may completely misinterpret it. Trying to get the output of the console app csproj lead me to the immediate failure. Yet heat yielded no error.

Can you send me a _Hello World_ style test case that demonstrates the use of Harvester?

Thank you,
Oleg

**Comment by GilesBathgate** [comment]
https://github.com/GilesBathgate/HarvesterExample

**Comment by oleg-shilo** [comment]
Excellent. It works. Probably was something in about the project structure in my case.

**Comment by oleg-shilo** [comment]
All done, in https://github.com/oleg-shilo/wixsharp/commit/a510ab80e9a2de314dae81d304d7430fe6ae55c0

Can you please have a look at `VsProject` class and update the XML doc for 
```
        public bool Binaries = true;
        public bool Content = true;
        public bool Satellites = true;
``` 

I am not entirely sure about meaning these fields in context of WiX Heat. You can do it as PR or even just respond to this message and I will update the code.

Thanks

**Comment by GilesBathgate** [comment]
>Can you please have a look at VsProject class and update the XML doc

These represent the project output groups harvested by heat:


**-pog:&lt;group&gt;** 
Specify output group of Visual Studio project, one of: Binaries, Symbols, Documents, Satellites, Sources, Content.

- Binaries - primary output of the project, e.g. the assembly exe or dll.
- Symbols - debug symbol files, e.g. pdb.
- Documents - documentation files.
- Satellites - the localized resource assemblies.
- Sources - source files.
- Content - content files.

This option may be repeated for multiple output groups; e.g. -pog:Binaries -pog:Content.


https://wixtoolset.org/documentation/manual/v3/overview/heat.html
